### DISCUSSION TITLE
Gradient Clipping in DDP/DDPSharded strategy behavior

### DISCUSSION THREAD

**Discussion by chetwinlow** [Lightning-AI/pytorch-lightning#19340] [DDP / multi-GPU / multi-node]
<img width="1608" alt="image" src="https://github.com/Lightning-AI/pytorch-lightning/assets/58224288/4d44a4e4-d7cf-4e30-8552-b72bb0c0b89e">

Based on my understanding, when using Pytorch DDP to do training (not lightning), loss.backward() automatically triggers a gradient sync / gradient averaging. However based on the provided image in one of the documentation, it says gradients are not synced until optimizer.step(), and the local effective batchsize is (bs * n_accums). If I have gradient clipping enabled, does it mean the gradient clipping is done before the sync / averaging or after the sync / average. I believe the correct approach is to clip after the global averaging / sync but I cannot find the actual implementation of this mechanism in the codebase, can someone help confirm the actual implementation of this flow in pytorch lightining? or shed some light on this matter?
